File: 2variant/code.cpp
#include <iostream>
using namespace std;

double power(double x, int n) {
    if (n == 0) {
        return 1;                  // Базовый случай: любая степень нуля равна 1
    } else if (n % 2 == 0) {      // Проверка четности
        double y = power(x, n / 2); // Рассчитываем половину степени
        return y * y;              // Возвращаем квадрат получившегося значения
    } else {
        double y = power(x, n / 2); // Нечётный случай
        return y * y * x;          // Добавочное умножение на x
    }
}

File: 2variant/code.java
public class PowerCalculator {
    public static double power(double x, int n) {
        if (n == 0) {
            return 1.0;                  // Базовый случай: любое число в степени 0 равно 1
        }
        if (n < 0) {
            return 1.0 / power(x, -n);  // Отрицательная степень: обратное значение
        }
        if (n % 2 == 0) {
            return power(x * x, n / 2); // Для четной степени: (x^2)^(n/2)
        }
        return x * power(x, n - 1);     // Для нечетной степени: x * x^(n-1)
    }

    public static void main(String[] args) {
        double base = 2.0;
        int exponent = 10;
        System.out.println(base + "^" + exponent + " = " + power(base, exponent));
    }
}
// Результат: 2.0^10 = 1024.0
